lib/remap-lang/src/expression/assignment.rs

```diff
@@ -1,19 +1,41 @@
 use super::Error as E;
 use crate::{
     expression::{Path, Variable},
-    state, Expr, Expression, Object, Result, TypeDef, Value,
+    state,
+    value::Kind,
+    Expr, Expression, Object, Result, TypeDef, Value,
 };
 
 #[derive(thiserror::Error, Clone, Debug, PartialEq)]
 pub enum Error {
     #[error("unable to insert value in path: {0}")]
     PathInsertion(String),
+
+    #[error(
+        r#"the {0} "{1}" does not need to handle the error-case, because its result is infallible"#
+    )]
+    UneededInfallibleAssignment(&'static str, String),
 }
 
 #[derive(Debug, Clone, PartialEq)]
 pub enum Target {
     Path(Path),
     Variable(Variable),
+    Infallible { ok: Box<Target>, err: Box<Target> },
+}
```

Comment:
I think we could take fuller advantage of Rusts type system here to [make illegal states unrepresentable](https://fsharpforfunandprofit.com/posts/designing-with-types-making-illegal-states-unrepresentable/).

```rust
pub enum Target {
  Path(Path),
  Variable(Variable)
}

pub enum AssignmentTarget {
  Fallible { ok: Box<Target> },
  Infallible { ok: Box<Target>, err: Box<Target> }
}
```

This makes it logically impossible to specify nested Infallible targets. It is much clearer what this code is expecting. Any bugs there will just be caught quickly by the compiler and we won't need all the `unimplemented!` calls below.

Reply:
I thought about doing that, but I didn't see the value in this specific case (I'm not averse to using `unimplemented`, since it's covered by the initializer and the parser, so if you ever change this, many tests would fail).

Would you be okay with me filling this as an issue for a future improvement, so that I can keep the pace up for 0.12 (this also requires some re-jiggling in the parser and module exporting, etc)?

Reply:
Sure!

Reply:
Thanks, tracking in https://github.com/timberio/vector/issues/5940.